### PR TITLE
fix silent failures in dispatch loop from stalling the pipeline

### PR DESCRIPTION
**Pull request by m-trieu**
fix silent failures in dispatch loop from stalling the pipeline

changed all executors used by the Streaming harness to use `TerminatingExecutors` which attaches a `WorkerUncaughtExceptionHandler` to all of the thread factories used to create an `Executor`/`ExecutorService`.  This means that any uncaught exception will terminate the JVM.  This seems to be the original intention in `DataflowWorkerHarnessHelper.initializeLogging()` since it attaches an uncaught exception handler that will terminate the JVM on uncaught exceptions, however that was only applied to raw threads.

added test in `SingleSourceWorkerHarness` to ensure exception is propogated to main thread.

`WorkerUncaughtExceptionHandler` already has a unit test to ensure that it works.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by m-trieu** [comment]
R: @arunpandianp @scwhittle

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/TerminatingExecutors.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  public static ExecutorService newFixedThreadPool(
+      int numThreads, ThreadFactoryBuilder threadFactoryBuilder, Logger logger) {
+    return Executors.newFixedThreadPool(
+        numThreads, terminatingThreadFactory(threadFactoryBuilder, logger));
+  }
+
+  private static ThreadFactory terminatingThreadFactory(
+      ThreadFactoryBuilder threadFactoryBuilder, Logger logger) {
+    return threadFactoryBuilder
+        .setUncaughtExceptionHandler(new WorkerUncaughtExceptionHandler(logger))
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/TerminatingExecutors.java:64]: I'm not sure this is working if the executor thread is catchign the exception and putting it in the future.

I think we might need instead to wrap the executor so that scheduled tasks are wrapped and then catch exceptions and call termination method like com/google/common/util/concurrent/WrappingExecutorService.java

**Review by scwhittle — CHANGES_REQUESTED** [review]

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/SingleSourceWorkerHarness.java [hunk 2]
```diff
@@ -103,11 +106,22 @@ public void start() {
...
+    while (isRunning.get()) {
+      Future<?> dispatchLoopFuture =
+          workProviderExecutor.submit(
+              () -> {
+                getDispatchLoop().run();
+                LOG.info("Dispatch done");
+              });
+
+      try {
+        dispatchLoopFuture.get();
```
[the remark this answers is not in the document]
Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/SingleSourceWorkerHarness.java:118]: removed

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/TerminatingExecutors.java (hunk 1 above)
Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/TerminatingExecutors.java:64]: added to test to ensure that we are halting the JVM

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/streaming/harness/SingleSourceWorkerHarnessTest.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        .build();
+  }
+
+  @Test
+  public void testDispatchLoopFailureThrowsException_appliance() {
+    RuntimeException expected = new RuntimeException("something bad happened");
+    SingleSourceWorkerHarness.GetWorkSender getWorkSender =
+        SingleSourceWorkerHarness.GetWorkSender.forAppliance(
+            () -> {
+              throw expected;
```
[the remark this answers is not in the document]
Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/streaming/harness/SingleSourceWorkerHarnessTest.java:64]: added test to ensure that the runtime is being terminated

**Comment by m-trieu** [comment]
back to you thank you!
@scwhittle

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/config/StreamingEngineComputationConfigFetcher.java [hunk 4]
```diff
@@ -157,6 +157,19 @@ private static Optional<StreamingConfigTask> fetchConfigWithRetry(
     }
   }
 
+  private static Optional<ComputationConfig> createComputationConfig(StreamingConfigTask config) {
```
Comment (arunpandianp) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/config/StreamingEngineComputationConfigFetcher.java:160]: can we revert the unrelated change, to keep the diff small and on topic?

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/StreamingWorkerStatusReporter.java [hunk 5]
```diff
@@ -228,6 +228,22 @@ private static void shutdownExecutor(ScheduledExecutorService executor) {
     }
   }
 
+  // Calculates the PerWorkerMetrics reporting frequency, ensuring alignment with the
```
Comment (arunpandianp) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/StreamingWorkerStatusReporter.java:231]: can we revert the unrelated change, to keep the diff small and on topic?

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 6]
```diff
@@ -346,10 +348,7 @@ public static StreamingDataflowWorker fromOptions(DataflowWorkerHarnessOptions o
             .setSizeMb(options.getWorkerCacheMb())
             .setSupportMapViaMultimap(options.isEnableStreamingEngine())
             .build();
-    Function<String, ScheduledExecutorService> executorSupplier =
-        threadName ->
-            Executors.newSingleThreadScheduledExecutor(
```
Comment (arunpandianp) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:351]: I think changing `Executors.newSingleThreadScheduledExecutor` to  `Executors.newSingleThreadExecutor` should fix the problem? We don't need to introduce a new Executor.

Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:351]: We need `ScheduledExecutorService` here and `Executors.newSingleThreadExecutor` does not return a `ScheduledExecutorService`

Question here:

do we want active work refresh failures to crash the JVM? if yes than we need a terminating executor, if not i think we can remove `TerminatingExecutors`, use  `Executors.newSingleThreadExecutor` everywhere we want a failure to crash the JVM and `ScheduledExecutorService` everywhere we are ok with failures.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/config/StreamingEngineComputationConfigFetcher.java (hunk 4 above)
Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/config/StreamingEngineComputationConfigFetcher.java:160]: done

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/StreamingWorkerStatusReporter.java (hunk 5 above)
Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/StreamingWorkerStatusReporter.java:231]: done

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/TerminatingExecutors.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+      int numThreads, ThreadFactoryBuilder threadFactoryBuilder, Logger logger) {
+    return new TerminatingExecutorService(
+        Executors.newFixedThreadPool(
+            numThreads, terminatingThreadFactory(threadFactoryBuilder, logger)));
+  }
+
+  public static TerminatingExecutorService newSingleThreadedExecutorForTesting(
+      JvmRuntime jvmRuntime, ThreadFactoryBuilder threadFactoryBuilder, Logger logger) {
+    return new TerminatingExecutorService(
+        Executors.newSingleThreadExecutor(
```
Comment (arunpandianp) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/TerminatingExecutors.java:77]: I think the test is passing since you have `newSingleThreadExecutor` here. Try changing this to `newSingleThreadScheduledExecutor`, the test will fail.

something like WrappingExecutorService suggested by @scwhittle seems better. We can use the new executor only where ScheduledExecutorService are needed (eg: refresh), other places we can use plain `Executors.newSingleThreadExecutor`

Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/TerminatingExecutors.java:77]: sgtm will update

Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/TerminatingExecutors.java:77]: do we want to crash the JVM on failures of tasks submitted to ScheduledExecutorService?

Comment (scwhittle) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/TerminatingExecutors.java:77]: I think it depends on the use case. There can be cases where the futures are being examined and having the exceptions manually handled by retrying etc.  If we are using schedule and not ignoring the future result then that seems like a case we wouldn't want a failure.  If we are using execute or otherwise ignoring the future it seems like we would want a failure.

I think Arun's advice to use ExecutorService where we can makes sense.  We can look at the remaining scheduledexecutorservice cases to see if we want to fail or if the errors are already handled via examining the futures.

Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/TerminatingExecutors.java]: updated thanks!

**Comment by m-trieu** [comment]
back to you @arunpandianp @scwhittle thanks!

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 6 above)
Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:351]: Resolved https://github.com/apache/beam/pull/32922#discussion_r1820519685

Comment (scwhittle) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:351]: yeah, since that schedules at a fixed rate internally to ActiveWorkRefresher I don think we want to crash the vm.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 8]
```diff
@@ -174,7 +174,7 @@ private StreamingDataflowWorker(
       StreamingCounters streamingCounters,
       MemoryMonitor memoryMonitor,
       GrpcWindmillStreamFactory windmillStreamFactory,
-      Function<String, ScheduledExecutorService> executorSupplier,
+      Function<String, ScheduledExecutorService> activeWorkRefreshExecutorFn,
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:177]: does this need to be a parameter still? can we just create a scheduled executor to pass into ActiveWorkRefresher?

**Review by scwhittle — CHANGES_REQUESTED** [review]

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/streaming/harness/SingleSourceWorkerHarnessTest.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+  private final GetDataClient getDataClient = mock(GetDataClient.class);
+  private final HeartbeatSender heartbeatSender = mock(HeartbeatSender.class);
+  private final Runnable waitForResources = () -> {};
+  private final Function<String, Optional<ComputationState>> computationStateFetcher =
+      ignored -> Optional.empty();
+  private final StreamingWorkScheduler streamingWorkScheduler = mock(StreamingWorkScheduler.class);
+
+  private SingleSourceWorkerHarness createWorkerHarness(
+      SingleSourceWorkerHarness.GetWorkSender getWorkSender, JvmRuntime runtime) {
+    Thread.setDefaultUncaughtExceptionHandler(new WorkerUncaughtExceptionHandler(runtime, LOG));
```
[the remark this answers is not in the document]
Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/streaming/harness/SingleSourceWorkerHarnessTest.java:54]: done

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 8 above)
Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:177]: done

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 10]
```diff
@@ -416,7 +413,8 @@ public static StreamingDataflowWorker fromOptions(DataflowWorkerHarnessOptions o
         streamingCounters,
         memoryMonitor,
         configFetcherComputationStateCacheAndWindmillClient.windmillStreamFactory(),
-        executorSupplier,
+        Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder().setNameFormat("threadName").build()),
```
Comment (arunpandianp) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:417]:
```suggestion
            new ThreadFactoryBuilder().setNameFormat("RefreshWork").build()),
```

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/streaming/harness/SingleSourceWorkerHarnessTest.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  private final GetDataClient getDataClient = mock(GetDataClient.class);
+  private final HeartbeatSender heartbeatSender = mock(HeartbeatSender.class);
+  private final Runnable waitForResources = () -> {};
+  private final Function<String, Optional<ComputationState>> computationStateFetcher =
+      ignored -> Optional.empty();
+  private final StreamingWorkScheduler streamingWorkScheduler = mock(StreamingWorkScheduler.class);
+
+  private SingleSourceWorkerHarness createWorkerHarness(
+      SingleSourceWorkerHarness.GetWorkSender getWorkSender, JvmRuntime runtime) {
+    // In non-test scenario this is set in StreamingDataflowWorker.main(...).
```
Comment (arunpandianp) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/streaming/harness/SingleSourceWorkerHarnessTest.java:54]:
```suggestion
    // In non-test scenario this is set in DataflowWorkerHarnessHelper.initializeLogging(...).
```

**Comment by arunpandianp** [comment]
lgtm

**Review by scwhittle — APPROVED** [review]
